third/nmh/man/refile.man
.\"
.\" %nmhwarning%
.\" $Id: refile.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH REFILE %manext1% MH.6.8 [%nmhversion%]
.SH NAME
refile \- file message in other folders
.SH SYNOPSIS
.in +.5i
.ti -.5i
refile 
\%[msgs] 
\%[\-draft]
\%[\-link] \%[\-nolink]
.br
\%[\-preserve] \%[\-nopreserve]
\%[\-unlink] \%[\-nounlink]
.br
\%[\-src\ +folder] 
\%[\-file\ file]
\%[\-rmmproc program]
.br
\%[\-normmproc]
+folder1 ...
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIRefile\fR moves (\fImv\fR\0(1)) or links (\fIln\fR\0(1)) messages
from a source folder into one or more destination folders.

If you think of a message as a sheet of paper, this operation is not
unlike filing the sheet of paper (or copies) in file cabinet folders.
When a message is filed, it is linked into the destination folder(s)
if possible, and is copied otherwise.  As long as the destination
folders are all on the same file system, multiple filing causes little
storage overhead.  This facility provides a good way to cross\-file or
multiply\-index messages.  For example, if a message is received from
Jones about the ARPA Map Project, the command

     refile\0cur\0+jones\0+Map

would allow the message to be found in either of the two folders `jones'
or `Map'.

You may specify the source folder using `\-src\ +folder'.  If this is
not given, the current folder is used by default.  If no message is
specified, then `cur' is used by default.

The option `\-file\ file' directs \fIrefile\fR to use the specified file
as the source message to be filed, rather than a message from a folder.
Note that the file should be a validly formatted message, just like
any other \fInmh\fR message.  It should \fBNOT\fR be in mail drop format
(to convert a file in mail drop format to a folder of \fInmh\fR messages,
see \fIinc\fR\0(1)).

If a destination folder doesn't exist, \fIrefile\fR will ask if you want
to create it.  A negative response will abort the file operation.  If the
standard input for \fIrefile\fR is \fInot\fR a tty, then \fIrefile\fR
will not ask any questions and will proceed as if the user answered
\*(lqyes\*(rq to all questions.

The option `\-link' preserves the source folder copy of the message (i.e.,
it does a \fIln\fR(1) rather than a \fImv\fR(1)), whereas, `\-nolink'
(the default) deletes the filed messages from the source folder.

Normally when a message is refiled, for each destination folder it
is assigned the number which is one above the current highest message
number in that folder.  Use of the `\-preserve' switch will override
this message renaming, and try to preserve the number of the message.
If a conflict for a particular folder occurs when using the `\-preserve'
switch, then \fIrefile\fR will use the next available message number
which is above the message number you wish to preserve.

If `\-link' is not specified (or `\-nolink' is specified), the filed
messages will be removed from the source folder.  The default is to
remove these messages by renaming them with a site-dependent prefix
(usually a comma).  Such files will then need to be removed in some
manner after a certain amount of time.  Many sites arrange for
\fIcron\fR\0(8) to remove these files once a day, so check with your
system administrator.

Alternately, if you wish for \fIrefile\fR to really remove the files
representing these messages from the source folder, you can use the
`-unlink' switch (not to be confused with the -link switch).  But
messages removed by this method cannot be later recovered.

.ne 4
If you prefer a more sophisticated method of `removing' the messages
from the source folder, you can define the \fIrmmproc\fR profile
component.  For example, you can add a profile component such as

	rmmproc:	/home/coleman/bin/rmm_msgs

then \fIrefile\fR will instead call the named program or script to
handle the message files.

The user may specify `\-rmmproc program' on the command line to
override this profile specification.  The `-normmproc' option forces
the message files to be deleted by renaming or unlinking them as
described above.

The `\-draft' switch tells \fIrefile\fR to file the <mh\-dir>/draft.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Folder\-Protect:~^To set mode when creating a new folder
.Ps
^rmmproc:~^Program to delete the message
.Sa
folder(1), rmf(1), rmm(1)
.De
`\-src\ +folder' defaults to the current folder
.Ds
`msgs' defaults to cur
.Ds
`\-nolink'
.Ds
`\-nounlink'
.Ds
`\-nopreserve'
.Co
If `\-src\ +folder' is given, it will become the current folder.
If neither `\-link' nor `all' is specified, the current message in the
source folder will be set to the last message specified; otherwise, the
current message won't be changed.

If the Previous\-Sequence profile entry is set, in addition to defining
the named sequences from the source folder, \fIrefile\fR will also define
those sequences for the destination folders.  See \fImh\-sequence\fR\0(5)
for information concerning the previous sequence.
.Bu
Since \fIrefile\fR uses your \fIrmmproc\fP to delete the message,
the \fIrmmproc\fP must \fBNOT\fP call \fIrefile\fP without specifying
`\-normmproc', or you will create an infinite loop.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/mh/uip/refile.c
/* refile.c - file messages away */
#ifndef	lint
static char ident[] = "@(#)$Id: refile.c,v 1.1.1.1 1996-10-07 07:14:20 ghudson Exp $";
#endif	/* lint */

#include "../h/mh.h"
#include <errno.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifdef LOCALE
#include	<locale.h>
#endif

/*  */

static struct swit switches[] = {
#define	DRAFTSW	0
    "draft", 0,

#define	LINKSW	1
    "link", 0,
#define	NLINKSW	2
    "nolink", 0,

#define	PRESSW	3
    "preserve", 0,
#define	NPRESSW	4
    "nopreserve", 0,

#define	SRCSW	5
    "src +folder", 0,

#define	FILESW	6
    "file file", 0,

#define	RPROCSW	7
    "rmmproc program", 0,
#define	NRPRCSW 8
    "normmproc", 0,

#define	HELPSW	9
    "help", 4,

    NULL, 0
};

/*  */

extern int  errno;


static char maildir[BUFSIZ];


struct st_fold {
    char   *f_name;
    struct msgs *f_mp;
};

static opnfolds(), clsfolds(), removeit();
/*  */

/* ARGSUSED */

main (argc, argv)
int     argc;
char  **argv;
{
    int	    linkf = 0,
            prsrvf = 0,
	    filep = 0,
            foldp = 0,
            msgp = 0,
	    isdf = 0,
	    i,
            msgnum;
    char   *cp,
           *folder = NULL,
            buf[100],
          **ap,
          **argp,
           *arguments[MAXARGS],
	   *filevec[NFOLDERS + 2],
          **files = &filevec[1],	/* leave room for removeit:vec[0] */
           *msgs[MAXARGS];
    struct st_fold   folders[NFOLDERS + 1];
    struct msgs *mp;

#ifdef LOCALE
	setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');
    if ((cp = m_find (invo_name)) != NULL) {
	ap = brkstring (cp = getcpy (cp), " ", "\n");
	ap = copyip (ap, arguments);
    }
    else
	ap = arguments;
    (void) copyip (argv + 1, ap);
    argp = arguments;

/*  */

    while (cp = *argp++) {
	if (*cp == '-')
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULLCP, "-%s unknown\n", cp);
		case HELPSW: 
		    (void) sprintf (buf, "%s [msgs] [switches] +folder ...",
			    invo_name);
		    help (buf, switches);
		    done (1);

		case LINKSW: 
		    linkf++;
		    continue;
		case NLINKSW: 
		    linkf = 0;
		    continue;

		case PRESSW: 
		    prsrvf++;
		    continue;
		case NPRESSW: 
		    prsrvf = 0;
		    continue;

		case SRCSW: 
		    if (folder)
			adios (NULLCP, "only one source folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULLCP, "missing argument to %s", argp[-2]);
		    folder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				   *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DRAFTSW:
		    if (filep > NFOLDERS)
			adios (NULLCP, "only %d files allowed!", NFOLDERS);
		    isdf = 0;
		    files[filep++] = getcpy (m_draft (NULLCP, NULLCP, 1, &isdf));
		    continue;
		case FILESW: 
		    if (filep > NFOLDERS)
			adios (NULLCP, "only %d files allowed!", NFOLDERS);
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULLCP, "missing argument to %s", argp[-2]);
		    files[filep++] = path (cp, TFILE);
		    continue;

		case RPROCSW: 
		    if (!(rmmproc = *argp++) || *rmmproc == '-')
			adios (NULLCP, "missing argument to %s", argp[-2]);
		    continue;
		case NRPRCSW: 
		    rmmproc = (char *)0;
		    continue;
	    }
	if (*cp == '+' || *cp == '@') {
	    if (foldp > NFOLDERS)
		adios (NULLCP, "only %d folders allowed!", NFOLDERS);
	    folders[foldp++].f_name =
		    path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	}
	else
	    msgs[msgp++] = cp;
    }

/*  */

    if (!m_find ("path"))
	free (path ("./", TFOLDER));
    if (foldp == 0)
	adios (NULLCP, "no folder specified");

#ifdef	WHATNOW
    if (!msgp && !foldp && !filep && (cp = getenv ("mhdraft")) && *cp)
	files[filep++] = cp;
#endif	/* WHATNOW */

    if (filep > 0) {
	if (folder || msgp)
	    adios (NULLCP, "use -file or some messages, not both");
	opnfolds (folders, foldp);
	for (i = 0; i < filep; i++)
	    if (m_file (files[i], folders, foldp, prsrvf))
		done (1);
	if (!linkf)
	    removeit (NULLMP, filep, filevec);
	done (0);
    }

    if (!msgp)
	msgs[msgp++] = "cur";
    if (!folder)
	folder = m_getfolder ();
    (void) strcpy (maildir, m_maildir (folder));

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");
    if (!(mp = m_gmsg (folder)))
	adios (NULLCP, "unable to read folder %s", folder);
    if (mp -> hghmsg == 0)
	adios (NULLCP, "no messages in %s", folder);

    for (msgnum = 0; msgnum < msgp; msgnum++)
	if (!m_convert (mp, msgs[msgnum]))
	    done (1);
    m_setseq (mp);

    opnfolds (folders, foldp);
    for (msgnum = mp -> lowsel; msgnum <= mp -> hghsel; msgnum++)
	if (mp -> msgstats[msgnum] & SELECTED) {
	    cp = getcpy (m_name (msgnum));
	    if (m_file (cp, folders, foldp, prsrvf))
		done (1);
	    free (cp);
	    if (!linkf) {
#ifdef	notdef
		mp -> msgstats[msgnum] |= DELETED;
#endif	/* notdef */
		mp -> msgstats[msgnum] &= ~EXISTS;
	    }
	}
    if (!linkf)
	mp -> msgflags |= SEQMOD;
    clsfolds (folders, foldp);

    m_replace (pfolder, folder);
    if (mp -> hghsel != mp -> curmsg
	    && (mp -> numsel != mp -> nummsg || linkf))
	m_setcur (mp, mp -> hghsel);
    m_sync (mp);
    m_update ();

    if (!linkf)
	removeit (mp, filep, filevec);

    done (0);
}

/*  */

static	opnfolds (folders, nfolders)
register struct st_fold *folders;
int	nfolders;
{
    register char  *cp;
    char    nmaildir[BUFSIZ];
    register struct st_fold *fp,
                            *ep;
    register struct msgs   *mp;
    struct stat st;

    for (ep = (fp = folders) + nfolders; fp < ep; fp++) {
	(void) chdir (m_maildir (""));
	(void) strcpy (nmaildir, m_maildir (fp -> f_name));

	if (stat (nmaildir, &st) == NOTOK) {
	    if (errno != ENOENT)
		adios (nmaildir, "error on folder");
	    cp = concat ("Create folder \"", nmaildir, "\"? ", NULLCP);
	    if (!getanswer (cp))
		done (1);
	    free (cp);
	    if (!makedir (nmaildir))
		adios (NULLCP, "unable to create folder %s", nmaildir);
	}

	if (chdir (nmaildir) == NOTOK)
	    adios (nmaildir, "unable to change directory to");
	if (!(mp = m_gmsg (fp -> f_name)))
	    adios (NULLCP, "unable to read folder %s", fp -> f_name);
	mp -> curmsg = 0;

	fp -> f_mp = mp;

	(void) chdir (maildir);
    }
}

/*  */

static	clsfolds (folders, nfolders)
register struct st_fold *folders;
int	nfolders;
{
    register struct st_fold *fp,
                           *ep;
    register struct msgs   *mp;

    for (ep = (fp = folders) + nfolders; fp < ep; fp++) {
	mp = fp -> f_mp;
	m_setseq (mp);
	m_sync (mp);
    }
}

/*  */

static	removeit (mp, filep, files)
register struct msgs *mp;
register int filep;
register char **files;
{
    register int    i,
                    vecp;
    register char  *cp,
                  **vec;

    if (rmmproc) {
	if (filep > 0) {
	    vec = files++;	/* filevec[1] */
	    files[filep] = NULL;
	}
	else {
	    if (mp -> numsel > MAXARGS - 2)
		adios (NULLCP, "more than %d messages for %s exec",
			MAXARGS - 2, rmmproc);
	    vec = (char **) calloc ((unsigned) (mp -> numsel + 2), sizeof *vec);
	    if (vec == NULL)
		adios (NULLCP, "unable to allocate exec vector");
	    vecp = 1;
	    for (i = mp -> lowsel; i <= mp -> hghsel; i++)
		if (mp -> msgstats[i] & SELECTED)
		    vec[vecp++] = getcpy (m_name (i));
	    vec[vecp] = NULL;
	}

	(void) fflush (stdout);
	vec[0] = r1bindex (rmmproc, '/');
	execvp (rmmproc, vec);
	adios (rmmproc, "unable to exec");
    }

    if (filep > 0) {
	files++;	/* filevec[1] */
	for (i = 0; i < filep; i++)
	    if (unlink (files[i]) == NOTOK)
		admonish (files[i], "unable to unlink");
    }
    else
	for (i = mp -> lowsel; i <= mp -> hghsel; i++)
	    if (mp -> msgstats[i] & SELECTED)
		if (unlink (cp = m_name (i)) == NOTOK)
		    admonish (cp, "unable to unlink");
}

/*  */

m_file (msg, folders, nfolders, prsrvf)
register char  *msg;
struct st_fold  *folders;
int	nfolders,
	prsrvf;
{
    int     in,
            out,
            linkerr,
            msgnum;
    register char  *nmsg;
    char    newmsg[BUFSIZ];
    register struct st_fold *fp,
			    *ep;
    register struct msgs *mp;
    struct stat st,
                s1;

    for (ep = (fp = folders) + nfolders; fp < ep; fp++) {
	mp = fp -> f_mp;
	if (prsrvf && (msgnum = m_atoi (nmsg = msg)) > 0) {
	    if (msgnum >= mp -> hghoff)
		if (mp = m_remsg (mp, 0, msgnum + MAXFOLDER))
		    fp -> f_mp = mp;
		else
		    adios (NULLCP, "unable to allocate folder storage");
	    if (!(mp -> msgstats[msgnum] & EXISTS)) {
		mp -> msgstats[msgnum] |= EXISTS;
#ifdef	notdef
		mp -> msgstats[msgnum] &= ~DELETED;
#endif	/* notdef */
		mp -> nummsg++;
	    }
	    mp -> msgstats[msgnum] |= SELECTED;		    
	    if (msgnum > mp -> hghmsg)
		mp -> hghmsg = msgnum;
	}
	else {
	    if (mp -> hghmsg >= mp -> hghoff)
		if (mp = m_remsg (mp, 0, mp -> hghoff + MAXFOLDER))
		    fp -> f_mp = mp;
		else
		    adios (NULLCP, "unable to allocate folder storage");

	    nmsg = m_name (msgnum = ++mp -> hghmsg);
	    mp -> nummsg++;
	    mp -> msgstats[msgnum] |= EXISTS | SELECTED;
	}
	if (mp -> lowmsg == 0)
	    mp -> lowmsg = msgnum;
	if (mp -> lowsel == 0 || msgnum < mp -> lowsel)
	    mp -> lowsel = msgnum;
	if (msgnum > mp -> hghsel)
	    mp -> hghsel = msgnum;

/*  */

	(void) sprintf (newmsg, "%s/%s", mp -> foldpath, nmsg);
	if (link (msg, newmsg) == NOTOK) {
#ifndef	EISREMOTE
	    linkerr = errno;
#else	/* EISREMOTE */
	    if ((linkerr = errno) == EISREMOTE)
		linkerr = EXDEV;
#endif	/* EISREMOTE */
	    if (linkerr == EEXIST
		    || (linkerr == EXDEV && stat (newmsg, &st) != NOTOK)) {
		if (linkerr != EEXIST
			|| stat (msg, &s1) == NOTOK
			|| stat (newmsg, &st) == NOTOK
			|| s1.st_ino != st.st_ino) {
		    advise (NULLCP, "message %s:%s already exists",
			    fp -> f_name, newmsg);
		    return 1;
		}
		continue;
	    }
	    if (linkerr == EXDEV) {
		if ((in = open (msg, 0)) == NOTOK) {
		    advise (msg, "unable to open message %s");
		    return 1;
		}
		(void) fstat (in, &st);
		if ((out = creat (newmsg, (int) st.st_mode & 0777))
			== NOTOK) {
		    advise (newmsg, "unable to create");
		    (void) close (in);
		    return 1;
		}
		cpydata (in, out, msg, newmsg);
		(void) close (in);
		(void) close (out);
	    }
	    else {
		advise (newmsg, "error linking %s to", msg);
		return 1;
	    }
	}
    }

    return 0;
}
